### PR TITLE
AIP-72: Allow pushing and pulling XCom from Task Context

### PR DESCRIPTION
**Pull request by kaxil**
Part of https://github.com/apache/airflow/issues/44481

There is a lot of cleanup to do but I wanted to get a basic DAG that uses XCom working first.

Example DAG used: `tutorial_dag`

```py

from __future__ import annotations

# [START tutorial]
# [START import_module]
import json
import textwrap

import pendulum

# The DAG object; we'll need this to instantiate a DAG
from airflow.models.dag import DAG

# Operators; we need this to operate!
from airflow.providers.standard.operators.python import PythonOperator

# [END import_module]

# [START instantiate_dag]
with DAG(
    "tutorial_dag",
    # [START default_args]
    # These args will get passed on to each operator
    # You can override them on a per-task basis during operator initialization
    default_args={"retries": 2},
    # [END default_args]
    description="DAG tutorial",
    schedule=None,
    start_date=pendulum.datetime(2021, 1, 1, tz="UTC"),
    catchup=False,
    tags=["example"],
) as dag:
    # [END instantiate_dag]
    # [START documentation]
    dag.doc_md = __doc__
    # [END documentation]

    # [START extract_function]
    def extract(**kwargs):
        ti = kwargs["ti"]
        data_string = '{"1001": 301.27, "1002": 433.21, "1003": 502.22}'
        ti.xcom_push("order_data", data_string)

    # [END extract_function]

    # [START transform_function]
    def transform(**kwargs):
        ti = kwargs["ti"]
        extract_data_string = ti.xcom_pull(task_ids="extract", key="order_data")
        order_data = json.loads(extract_data_string)

        total_order_value = 0
        for value in order_data.values():
            total_order_value += value

        total_value = {"total_order_value": total_order_value}
        total_value_json_string = json.dumps(total_value)
        ti.xcom_push("total_order_value", total_value_json_string)

    # [END transform_function]

    # [START load_function]
    def load(**kwargs):
        ti = kwargs["ti"]
        total_value_string = ti.xcom_pull(task_ids="transform", key="total_order_value")
        total_order_value = json.loads(total_value_string)

        print(total_order_value)

    # [END load_function]

    # [START main_flow]
    extract_task = PythonOperator(
        task_id="extract",
        python_callable=extract,
    )
    extract_task.doc_md = textwrap.dedent(
        """\
    #### Extract task
    A simple Extract task to get data ready for the rest of the data pipeline.
    In this case, getting data is simulated by reading from a hardcoded JSON string.
    This data is then put into xcom, so that it can be processed by the next task.
    """
    )

    transform_task = PythonOperator(
        task_id="transform",
        python_callable=transform,
    )
    transform_task.doc_md = textwrap.dedent(
        """\
    #### Transform task
    A simple Transform task which takes in the collection of order data from xcom
    and computes the total order value.
    This computed value is then put into xcom, so that it can be processed by the next task.
    """
    )

    load_task = PythonOperator(
        task_id="load",
        python_callable=load,
    )
    load_task.doc_md = textwrap.dedent(
        """\
    #### Load task
    A simple Load task which takes in the result of the Transform task, by reading it
    from xcom and instead of saving it to end user review, just prints it out.
    """
    )

    extract_task >> transform_task >> load_task
```

---
image

image

---
image

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: tests/api_fastapi/execution_api/routes/test_xcoms.py [hunk 1]
```diff
@@ -40,19 +41,21 @@ class TestXComsGetEndpoint:
...
         ("value", "expected_value"),
         [
-            ("value1", "value1"),
-            ({"key2": "value2"}, {"key2": "value2"}),
-            ({"key2": "value2", "key3": ["value3"]}, {"key2": "value2", "key3": ["value3"]}),
-            (["value1"], ["value1"]),
+            ('"value1"', '"value1"'),
+            ('{"key2": "value2"}', '{"key2": "value2"}'),
+            ('{"key2": "value2", "key3": ["value3"]}', '{"key2": "value2", "key3": ["value3"]}'),
+            ('["value1"]', '["value1"]'),
```
Comment (kaxil) [inline tests/api_fastapi/execution_api/routes/test_xcoms.py:47]: Now the contract between API Server and its client is simple: It expects JSON serialized string and will return string. 

It pushes down responsibility to serialize it into native (Python, Golang) format to the respective clients.

In one of the future PRs, I will move some of the serialization and deserialization logic in Python clients

File: task_sdk/src/airflow/sdk/api/client.py [hunk 2]
```diff
@@ -207,11 +207,16 @@ class XComOperations:
     def __init__(self, client: Client):
         self.client = client
 
-    def get(self, dag_id: str, run_id: str, task_id: str, key: str, map_index: int = -1) -> XComResponse:
+    def get(
+        self, dag_id: str, run_id: str, task_id: str, key: str, map_index: int | None = -1
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/api/client.py]: If this can be none, shouldn't that be the default?

File: task_sdk/src/airflow/sdk/api/client.py [hunk 3]
```diff
@@ -207,11 +207,16 @@ class XComOperations:
...
+    ) -> XComResponse:
         """Get a XCom value from the API server."""
         # TODO: check if we need to use map_index as params in the uri
         # ref: https://github.com/apache/airflow/blob/v2-10-stable/airflow/api_connexion/openapi/v1.yaml#L1785C1-L1785C81
-        resp = self.client.get(f"xcoms/{dag_id}/{run_id}/{task_id}/{key}", params={"map_index": map_index})
+        params = {}
+        if map_index:
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/api/client.py]:
```suggestion
        if map_index is not None
```

Map index 0 is valid

File: task_sdk/src/airflow/sdk/execution_time/task_runner.py [hunk 4]
```diff
@@ -111,11 +115,109 @@ def get_template_context(self):
...
 
-    def xcom_pull(self, *args, **kwargs): ...
+    def xcom_pull(
+        self,
+        task_ids: str | None = None,  # TODO: Simplify to a single task_id? (breaking change)
+        dag_id: str | None = None,
+        key: str = "return_value",  # TODO: Make this a constant (``XCOM_RETURN_KEY``)
+        include_prior_dates: bool = False,  # TODO: Add support for this
+        *,
+        map_indexes: int | None = None,
```
Comment (ashb) [inline task_sdk/src/airflow/sdk/execution_time/task_runner.py]: Name says plural, but type only allows one value

**Review by ashb — APPROVED** [review]
A few nits

File: task_sdk/src/airflow/sdk/api/client.py (hunk 2 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/api/client.py]: True, resolved in https://github.com/apache/airflow/pull/45075/commits/ae097e8a3cf26c2d3f9df6f97895a29a07a08500

File: task_sdk/src/airflow/sdk/api/client.py (hunk 3 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/api/client.py]: Resolved in https://github.com/apache/airflow/pull/45075/commits/ae097e8a3cf26c2d3f9df6f97895a29a07a08500

File: task_sdk/src/airflow/sdk/execution_time/task_runner.py (hunk 4 above)
Comment (kaxil) [reply task_sdk/src/airflow/sdk/execution_time/task_runner.py]: Resolved in https://github.com/apache/airflow/pull/45075/commits/ae097e8a3cf26c2d3f9df6f97895a29a07a08500

**Pull request merged**